src/storage/src/storage_state.rs

```diff
@@ -1041,13 +1042,26 @@ impl<'w, A: Allocate> Worker<'w, A> {
                         }
                     }
                 }
-                StorageCommand::InitializationComplete
-                | StorageCommand::UpdateConfiguration(_)
-                | StorageCommand::CreateSinks(_) => (),
+                StorageCommand::CreateSinks(exports) => {
+                    // Ensure that exports are forward-rolling alter compatible.
+                    for export in exports {
+                        let prev = running_exports_descriptions
+                            .insert(export.id, export.description.clone());
+
+                        if let Some(prev_export) = prev {
+                            prev_export
+                                .alter_compatible(export.id, &export.description)
```

Comment:
This doesn't feel it belongs here. The cluster should process the commands sent by the controller without much questioning. Questions like validity/type checking/compatibility are planning concerns and should have been resolved already

Reply:
This is a descendant of [the initial reconciliation implementation in storage_state](https://github.com/MaterializeInc/materialize/pull/13526/files#diff-752224caae5760da0459a31cacb0fe5853e6c17de0b4027598e8da60f9c5f092R275-R282). Back then, it relied on strict equality, but that has obviously shifted.

You propose we rip out all of these guards?

Reply:
Hm, that's a good point. I'm not sure why this was put there in the first place so now I'm doubting myself. If this assertion fires will the new cluster get into the same piece of code and panic again or will the new cluster "skip ahead" and just get a consolidated command stream where the alteration is gone? If the latter then isn't it better to not panic in the first place and instead tear down the old dataflow and start the new one?

We don't have to resolve these questions in this PR as this turned out to be a bit more complex than I expected. Your initial change seems like the correct thing to do in the context of the current state of affairs

Reply:
A new cluster would not get into this state because it wouldn't have existing state to reconcile against (i.e. it starts anew). I'm glad to leave this be as you suggest and we can revisit.

Reply:
Sounds good, and yes if that's the case then this is an argument for tearing down the previous version and installing the new one since the after panic cluster will do the same thing. Definitely something to revisit in another PR though